pkgs/build-support/fetchgitlab/default.nix

```
@@ -1,32 +1,57 @@
-{ fetchgit, fetchzip, lib }:
+{ lib, fetchgit, fetchzip }:
 
 # gitlab example
 { owner, repo, rev, protocol ? "https", domain ? "gitlab.com", name ? "source", group ? null
-, fetchSubmodules ? false, leaveDotGit ? false, deepClone ? false
+, fetchSubmodules ? false, leaveDotGit ? false
+, deepClone ? false, forceFetchGit ? false
+, sparseCheckout ? ""
+, private ? false, varPrefix ? null
 , ... # For hash agility
 } @ args:
 
 let
   slug = lib.concatStringsSep "/" ((lib.optional (group != null) group) ++ [ owner repo ]);
   escapedSlug = lib.replaceStrings [ "." "/" ] [ "%2E" "%2F" ] slug;
   escapedRev = lib.replaceStrings [ "+" "%" "/" ] [ "%2B" "%25" "%2F" ] rev;
-  passthruAttrs = removeAttrs args [ "protocol" "domain" "owner" "group" "repo" "rev" ];
+  passthruAttrs = removeAttrs args [ "protocol" "domain" "owner" "group" "repo" "rev" "forceFetchGit" "private" "varPrefix" ];
 
-  useFetchGit = deepClone || fetchSubmodules || leaveDotGit;
+  varBase = "NIX${if varPrefix == null then "" else "_${varPrefix}"}_GITLAB_PRIVATE_";
+  useFetchGit = fetchSubmodules || leaveDotGit || deepClone || forceFetchGit || (sparseCheckout != "");
   fetcher = if useFetchGit then fetchgit else fetchzip;
 
+  privateAttrs = lib.optionalAttrs private (
+    lib.throwIfNot (protocol == "https")
+    "private token login is only supported for https" {
+      netrcPhase = ''
+        if [ -z "''$${varBase}USERNAME" -o -z "''$${varBase}PASSWORD" ]; then
+          echo "Error: Private fetchFromGitLab requires the nix building process (nix-daemon in multi user mode) to have the ${varBase}USERNAME and ${varBase}PASSWORD env vars set." >&2
+          exit 1
+        fi
+      '' + (if useFetchGit then ''
+        cat > netrc <<EOF
+        machine ${domain}
+                login ''$${varBase}USERNAME
+                password ''$${varBase}PASSWORD
+        EOF
+      '' else ''
+        touch netrc
+        curlOpts="$curlOpts --header PRIVATE-TOKEN:''$${varBase}PASSWORD"
```

In principle, there should be a space between the header name and argument, but this would require curlOpts to be an array as in #164662.

```suggestion
        curlOptsList+=("--header" "PRIVATE-TOKEN: ''$${varBase}PASSWORD")
```

---

Implemented with a temporary file and `@filename` syntax, which is anyway preferred to avoid leaking the token on the command line.